# Исправление опечаток в запросе

## Пример с лекции

Подберём к вопросу похожий на него вопрос.

In [ ]:
!pip install --upgrade --no-cache-dir gdown

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
  Created wheel for gdown: filename=gdown-4.4.0-py3-none-any.whl size=14774 sha256=6ba5a768d83c2ff11705ed7d554c6a9787bd920a78cb7340e667885caf314edb
  Stored in directory: /tmp/pip-ephem-wheel-cache-ddm_3l5l/wheels/fb/c3/0e/c4d8ff8bfcb0461afff199471449f642179b74968c15b7a69c
Successfully built gdown
  Attempting uninstall: gdown
    Found existing installation: gdown 4.2.2
    Uninstalling gdown-4.2.2:
      Successfully uninstalled gdown-4.2.2


In [ ]:
!git clone https://github.com/vladislavneon/RuBQ.git

Cloning into 'RuBQ'...
remote: Enumerating objects: 106, done.
remote: Counting objects: 100% (106/106), done.
remote: Compressing objects: 100% (89/89), done.
remote: Total 106 (delta 56), reused 29 (delta 15), pack-reused 0
Receiving objects: 100% (106/106), 14.13 MiB | 15.66 MiB/s, done.
Resolving deltas: 100% (56/56), done.


In [ ]:
import json

filepath = "/content/RuBQ/RuBQ_2.0/RuBQ_2.0_test.json"
with open(filepath, mode="r", encoding="utf-8") as input:
  lines = input.readlines()
res_json = json.loads(''.join(lines))

In [ ]:
dataset_temp = [x["question_text"] for x in res_json]

In [ ]:
dataset_temp[:20]

['Что может вызвать цунами?',
 'Кто написал роман «Хижина дяди Тома»?',
 'Кто автор пьесы «Ромео и Джульетта»?',
 'Как называется столица Румынии?',
 'На каком инструменте играл Джимми Хендрикс?',
 'Какой стране принадлежит остров Таити?',
 'Кто создал Зимний дворец в Санкт-Петербурге?',
 'Кто режиссер фильма "Бриллиантовая рука"?',
 'Какой стране принадлежат Канарские острова?',
 'Какой географический объект называется Лимпопо?',
 'Какое современное государство занимает территорию, на которой находился древний ассирийский город Ниневия?',
 'По какой шкале измеряют силу землетрясений?',
 'В какой стране впервые появилась игра в гольф?',
 'Каким видом спорта увлекался президент РФ В.В.Путин?',
 'Как звали отца Северуса Снейпа?',
 'Как звали сына Драко?',
 'Как звали маму Северуса Снейпа?',
 'Как звали создателя факультета Когтевран (Рейвенкло)?',
 'Как звали дракончика Хагрида?',
 "Кто автор исторического романа 'Княжна Тараканова'?"]

In [ ]:
len(dataset_temp)

2330

Тут всего 2330 вопросов и нет никакой информации о похожести. Для решения задачи нам понадобиться ещё один датасет.

Хороший способ скачивать с Google Drive

In [ ]:
!gdown --id --no-cookies 1l-EOL-wj_noSaIQ89qsS-Be0SQnBcLk4

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1l-EOL-wj_noSaIQ89qsS-Be0SQnBcLk4
To: /content/quora_question_pairs_rus.csv.zip
100% 26.7M/26.7M [00:00<00:00, 84.0MB/s]


In [ ]:
!unzip quora_question_pairs_rus.csv.zip

Archive:  quora_question_pairs_rus.csv.zip
  inflating: quora_question_pairs_rus.csv  


Это большой датасет вопросов quora, который был автоматически переведён.

In [ ]:
import pandas as pd;

In [ ]:
quora_df = pd.read_csv('quora_question_pairs_rus.csv', usecols=lambda x: 'Unnamed' not in x)

In [ ]:
quora_df.shape

(404288, 3)

In [ ]:
quora_df.sample(20)

question1  \
11546   какой худший медицинский совет вы когда-либо п...   
311452  в моей кнопке ноутбука c не работает, как я мо...   
341606                что происходит внутри камеры облака   
11589   который предложил narendra modi остановить цир...   
11576   является демонициализацией политики modi хорош...   
315454  где я могу найти онлайн-курс для цифрового мар...   
110275  какие инструменты используются для измерения з...   
73380   как он чувствует себя гомосексуалистом в Индон...   
367858  почему малайцы и сингапурцы относятся к индоне...   
369656  насколько велик рынок финансирования акций в С...   
118013  каковы различные этапы и химические вещества, ...   
325565  когда вы делаете онлайн-знакомства, а кто-то с...   
240863  каковы лучшие инструменты для создания вовлече...   
128317  веб-разработка растет и быстро развивается, я ...   
127662   сделаю 2-дневный раскол, сделаю меня мускулистым   
54917   как вы считаете сильным в арабском, если вы па...   
137241  какой модуль python является лучшим посланнико...   
18585             что делает хорошего английского учителя   
241673  почему я не могу глубоко задуматься, когда я г...   
282841  каково будет влияние на Индию, если коалиция Д...   

                                                question2  is_duplicate  
11546        как вы рассказали родителям о своей подруге?             0  
311452   как убеждение трансформируется по определенности             0  
341606     то, что может попасть в черную дыру, похоже на             0  
11589         почему запрет на замещение 500 и 1000 рупий             1  
11576   это демонизация денег, действительно работающа...             1  
315454  лучший курс для цифрового маркетинга онлайн дл...             1  
110275  какой инструмент используется для измерения зе...             1  
73380          как он чувствует себя привлекательным геем             0  
367858  каков статус моей собственности в Сингапуре, е...             0  
369656                          черный костюм лучший цвет             0  
118013        Мое полотенце может быть пляжным полотенцем             0  
325565  мы находимся в дальних отношениях более 3 меся...             0  
240863  существует ли такая веб-служба, как то, что я ...             0  
128317  я буду считаться ленивым со стороны работодате...             0  
127662  моя цель - сделать 1 миллион долларов в год, е...             0  
54917                  как вы говорите, мечеть в арабском             0  
137241                   у чего должны быть модули python             0  
18585       что делает хорошего учителя английского языка             1  
241673                почему я не могу глубоко задуматься             1  
282841  что влияет на Индию, если король Дональда стан...             1

In [ ]:
dataset_temp.extend(quora_df["question1"].values)

In [ ]:
dataset_temp.extend(quora_df["question2"].values)

In [ ]:
len(dataset_temp)

810906

Скачаем эмбеддинги от Natasha (https://natasha.github.io/navec/)

In [ ]:
!pip install navec

!wget https://storage.yandexcloud.net/natasha-navec/packs/navec_hudlit_v1_12B_500K_300d_100q.tar

--2022-03-13 16:16:35--  https://storage.yandexcloud.net/natasha-navec/packs/navec_hudlit_v1_12B_500K_300d_100q.tar
Resolving storage.yandexcloud.net (storage.yandexcloud.net)... 213.180.193.243, 2a02:6b8::1d9
Connecting to storage.yandexcloud.net (storage.yandexcloud.net)|213.180.193.243|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 53012480 (51M) [application/x-tar]
Saving to: ‘navec_hudlit_v1_12B_500K_300d_100q.tar’

navec_hudlit_v1_12B 100%[===================>]  50.56M  15.5MB/s    in 3.3s    

2022-03-13 16:16:39 (15.5 MB/s) - ‘navec_hudlit_v1_12B_500K_300d_100q.tar’ saved [53012480/53012480]



In [ ]:
from navec import Navec

path = 'navec_hudlit_v1_12B_500K_300d_100q.tar'
navec = Navec.load(path)

In [ ]:
navec.get('python')

array([ 0.19214402, -0.1959362 ,  0.43695402, -0.33588752,  0.39676988,
       -0.07734875,  0.04260164,  0.67186844, -0.04072431, -0.07451263,
        0.49101624,  0.07942662,  0.20798121,  0.18426852, -0.30046153,
       -0.15204567,  0.36486024, -0.85307485,  0.48137093,  0.427026  ,
       -0.6845362 , -0.08942661, -0.3101123 , -0.45371357, -0.611568  ,
       -0.26479366, -0.9259754 , -0.11309931, -0.48673463,  0.42229277,
        0.20572862, -0.13167205, -0.8779028 ,  0.07185546,  0.02660668,
       -0.0868782 ,  0.572537  ,  0.6779026 , -0.11222838,  0.14546263,
       -0.03367202,  0.24061641,  0.08572771,  0.00990655,  0.25097123,
        0.01056161, -0.51034373,  0.09052686,  0.43652472,  0.3318286 ,
       -0.11107466, -0.14714225, -0.34816197, -0.3292252 ,  0.33444104,
       -0.07015625,  0.00249229,  0.03096647, -0.39550957, -0.37711838,
       -0.15093368, -0.3094792 ,  0.17217222,  0.21431988, -0.6753277 ,
       -0.5604233 ,  0.46426848,  0.08588701,  0.44196838,  0.34

Тут нет удобных штук типа similarity

In [ ]:
navec.similarity('python', 'java')

AttributeError: ignored

In [ ]:
import numpy as np

In [ ]:
def cosine(u, v):
  res = np.dot(u, v) / (np.linalg.norm(u) * np.linalg.norm(v))
  return res

In [ ]:
print(cosine(navec['python'],navec['java']))

0.63120574


In [ ]:
print(cosine(navec['python'],navec['змея']))

-0.008952813


In [ ]:
print(cosine(navec['питон'],navec['змея']))

0.36701798


In [ ]:
print(cosine(navec['питон'],navec['java']))

0.030473152


In [ ]:
print(cosine(navec['индонезия'],navec['java']))

0.16179799


In [ ]:
print(cosine(navec['индонезия'],navec['ява']))

0.51982415


In [ ]:
from nltk.tokenize import WordPunctTokenizer

tokenizer = WordPunctTokenizer()

In [ ]:
dataset = dataset_temp.copy()

Строим эмбеддинги для предложений по эмбеддингам для слов

In [ ]:
def embed(text):
  res = []

  tokens = tokenizer.tokenize(str(text).lower())

  for token in tokens:
    if token in navec:
      res.append(navec[token])

  if not res:
    res.append(navec['<unk>']) # это из документации

  return np.mean(res, axis=0)

In [ ]:
embed("Тестовый текст - это просто текст для теста")

array([-0.09564245, -0.01535456,  0.12325782, -0.06663185, -0.20047495,
        0.03183787,  0.26239553,  0.35224754,  0.2521219 , -0.07307456,
       -0.307853  ,  0.01662676,  0.02816503,  0.06181916,  0.06616659,
       -0.08646315, -0.01595033, -0.18428268,  0.03257051,  0.30827984,
       -0.1424056 , -0.12214033, -0.06059988,  0.34464318,  0.06410901,
       -0.12486495, -0.33976468,  0.38941613, -0.05564322, -0.08336477,
       -0.00159295,  0.30793437,  0.5169112 ,  0.39479265, -0.09011184,
        0.08616064, -0.04342553, -0.01709452,  0.00687668,  0.04990578,
        0.09723997,  0.06334966,  0.0321274 , -0.07932533,  0.23986931,
       -0.08293863, -0.00717402,  0.05017046,  0.05389752, -0.15716644,
       -0.1395224 , -0.25135663, -0.22502461,  0.01293752,  0.05510895,
        0.2168825 , -0.16007565, -0.01202961, -0.2204808 ,  0.16102086,
        0.20584778, -0.39431804, -0.0983623 ,  0.00144196,  0.09941299,
        0.1484852 ,  0.24532807,  0.3359027 ,  0.13392086,  0.05

In [ ]:
embedded_data = [embed(x) for x in dataset] # эмбеддинги строятся примерно минуту

In [ ]:
def get_result(text):
  query = embed(text)
  cosines = [cosine(x, query) for x in embedded_data]

  for i in np.argpartition(cosines, -10)[-10:]: # Функция argpartition() возвращает массив индексов элементов исходного массива в их разбиении по указанному значению
    print(dataset[i])

In [ ]:
text = "Какой город самый большой в мире?"

get_result(text)

какой город в мире самый красивый, чтобы жить в
какой лучший город в мире
какой лучший город в мире
какой самый большой океан в мире
какой самый большой океан в мире
какой самый большой океан в мире
какой самый большой город в мире
какой самый большой город в мире
какой самый большой город в мире по площади
какой самый большой город в мире


In [ ]:
text = "Какой вуз самый лучший?"

get_result(text)

какой лучший университет
какой лучший университет
какой лучший колледж
какой лучший колледж
какой лучший dslr
какой лучший dslr
какой лучший колледж
какой лучший dslr
какой лучший метапоиска
какой лучший видеоредактор


In [ ]:
text = "Какой драматург написал пьесу «Ромео и Джульетта»"

get_result(text)

Кто написал оперу Евгений Онегин?
Уильям Шекспир писал что-нибудь, кроме пьес
Уильям Шекспир писал что-нибудь, кроме пьес
Какой композитор написал первую советскую рок-оперу «Орфей и Эвридика»?
как написать сценарий комедии
Кто автор пьесы «Ромео и Джульетта»?
что написал Уильям Шекспир помимо пьес
какие пьесы сыграл соавтор шекспира
Какую оперу на сюжет «Ромео и Джульетты» написал композитор Винченцо Беллини?
что написал Уильям Шекспир помимо пьес


In [ ]:
text = "Какие причины у глобального потепления?"

get_result(text)

какие убедительные доказательства существуют в отношении существования глобального потепления и изменения климата
какие возможные выгоды могут быть для глобального потепления
какие возможные выгоды могут быть для глобального потепления
каковы некоторые важные причины и следствия в отношении глобального потепления
каковы причины глобального потепления
каковы основные причины глобального потепления
каковы проблемы и последствия глобального потепления
какие последствия глобального потепления
каковы основные причины глобального потепления
каковы проблемы и последствия глобального потепления


In [ ]:
text = "Какой сериал можно посмотреть на выходных?"

get_result(text)

где я могу посмотреть фильм на море
какой фильм я должен смотреть на netflix
какой фильм я должен смотреть на netflix
какой болливудский фильм лучше всего посмотреть в эти выходные
какой фильм стоит посмотреть на rustom или mohenjo daro
лучше смотреть на фурго, когда сериал видел фильм первым или нет
что такое ужасный фильм, который можно посмотреть на Хэллоуин
что такое ужасный фильм, который можно посмотреть на Хэллоуин
где я могу скачать или посмотреть сериал серии krishi
какой фильм я должен смотреть на netflix


In [ ]:
text = "Кокой город самй бальшой?"

get_result(text)

Кем был основан город Нижний Новгород?
Бангалор - глобальный город
город Пуне, метро
Торонто опасный город
Мехико город опасен
Торонто - безопасный город
как город Бангалор
новый город Хэмптон
Сиэтл - хороший город
как город бангалор


## Упражнение 2

Из примера видно, что с опечатками плохо работает. Можно попробуем улучшить с помощью fasttext (см/ лекцию (https://colab.research.google.com/drive/1Tv6LPSEaYZeVt5_SjVlS_vmb86f9bp2I?usp=sharing) или с помощью исправления опечаток (см. лекцию https://colab.research.google.com/drive/1UHlWAOoWqHoEzCZuS7HhyF_xATPK8_Oq?usp=sharing).

In [ ]:
import requests
import string

Удалить пунктуацию

In [ ]:
def clear_punc(text):
  table = str.maketrans(dict.fromkeys(string.punctuation + '«»'))
  text = text.translate(table)
  return text

In [ ]:
print(clear_punc("Какой город самый большой?"))

Какой город самый большой


Убрать опечатки

In [ ]:
def check_word(word):
  domain = "https://speller.yandex.net/services/spellservice.json"
  request = requests.get(domain + "/checkText?text=" + word)
  if len(request.json()):
    return request.json()[0]["s"][0]
  else:
    return word

In [ ]:
def fix_query(text):
  words = text.split()
  fixed = ''
  for word in words:
    fixed = fixed + ' ' + check_word(word)
  return fixed

In [ ]:
test = "Кокой город самй бальшой?"
test_f = fix_query(test)
test_std = "Какой город самый большой?"

In [ ]:
print(test_f)

 Какой город самый большой


In [ ]:
a = get_result(test_f)

какой самый интересный красивый крутой город или город, который вы посетили
какой самый большой дом в skyrim
какой самый большой дом в skyrim
какой самый большой город в Финляндии?
какой самый большой город в мире
какой самый большой город в мире
какой самый большой город в мире
какой самый большой город
какой самый большой город
какой самый большой город


Возвращать только уникальные запросы

In [ ]:
def get_uniq_result(text):
  ans = set()
  query = embed(text)
  cosines = [cosine(x, query) for x in embedded_data]

  for i in np.argpartition(cosines, -10)[-10:]: # Функция argpartition() возвращает массив индексов элементов исходного массива в их разбиении по указанному значению
    ans.add(dataset[i])
  return ans

In [ ]:
b = get_uniq_result(test_f)

In [ ]:
print(b)

{'какой самый большой дом в skyrim', 'какой самый большой город в мире', 'какой самый большой город', 'какой самый большой город в Финляндии?', 'какой самый интересный красивый крутой город или город, который вы посетили'}


Итог

In [ ]:
def search(text):
  text = fix_query(clear_punc(text))
  text = get_uniq_result(text)
  return text

Тесты

In [ ]:
tests = []
tests.append('Какой город самый большой в мире?')
tests.append('Какой вуз самый лучший?')
tests.append('Какой драматург написал пьесу «Ромео и Джульетта»?')
tests.append('Какие причины у глобального потепления?')
tests.append('Какой сериал можно посмотреть на выходных?')
tests.append('как я зарабатываю деньги в Интернете')
tests.append('Как называется столица Румынии?')
tests.append('какие самые глупые вопросы задают на quora')
tests.append('Как звали дракончика Хагрида?е')
tests.append('лучший споЧоб избежать рака')

In [ ]:
print(tests)

['Какой город самый большой в мире?', 'Какой вуз самый лучший?', 'Какой драматург написал пьесу «Ромео и Джульетта»?', 'Какие причины у глобального потепления?', 'Какой сериал можно посмотреть на выходных?', 'как я зарабатываю деньги в Интернете', 'Как называется столица Румынии?', 'какие самые глупые вопросы задают на quora', 'Как звали дракончика Хагрида?е', 'лучший споЧоб избежать рака']


In [ ]:
for t in tests:
  print(t)
  print(search(t))
  print('________')

Какой город самый большой в мире?
{'какой самый большой город в мире', 'какой лучший город в мире', 'какой самый большой океан в мире', 'какой город в мире самый красивый, чтобы жить в', 'какой самый большой город в мире по площади'}
________
Какой вуз самый лучший?
{'какой лучший университет', 'какой лучший колледж', 'какой лучший метапоиска', 'какой лучший видеоредактор', 'какой лучший dslr'}
________
Какой драматург написал пьесу «Ромео и Джульетта»?
{'Какую оперу на сюжет «Ромео и Джульетты» написал композитор Винченцо Беллини?', 'как написать сценарий комедии', 'Уильям Шекспир писал что-нибудь, кроме пьес', 'Кто написал оперу Евгений Онегин?', 'Кто автор пьесы «Ромео и Джульетта»?', 'что написал Уильям Шекспир помимо пьес', 'Какой композитор написал первую советскую рок-оперу «Орфей и Эвридика»?', 'какие пьесы сыграл соавтор шекспира'}
________
Какие причины у глобального потепления?
{'какие убедительные доказательства существуют в отношении существования глобального потепления и 